### ISSUE TITLE
Support key/value records in the core language

### ISSUE THREAD

**Issue by haasn** [mesonbuild/meson#2463]
When building custom abstractions such as lists of dependencies that you intend to process via `foreach` loops (cf. #2454), it's a bit annoying to have to use arrays and indices to represent my data.

It would probably be a good idea to add support for a native key/value record, with syntax similar to the keyword arguments (and probably using the traditional `{ }`). Then I could write neat stuff like this:

```
components = [
  {
    name: 'foo',
    dep: dependency('foo', '>=1.0'),
    sources: [ 'foo.c', 'foo_utils.c' ]
  }, {
    ...
]
```

and process this with a foreach loop that builds up my `dependencies:` and `sources:` arrays based on the results of the configure check.

This same example actually provides a second motivating case for key/value records: suppose I want to extend this to also allow dependencies to `require:` other dependencies, and have them automatically skipped if the required other dependencies are unavailable.

In this case, the natural way of implementing it would be to set `found[c.name] = c.dep.found()` to store up a table of dependencies which we've already found, and then do a check against `found[r]` for each r : c.requires to see if all of the required dependencies have already been met.

**Comment by haasn** [comment]
> This same example actually provides a second motivating case for key/value records: suppose I want to extend this to also allow dependencies to require: other dependencies, and have them automatically skipped if the required other dependencies are unavailable.

It actually seems that this particular use case is already covered by `configuration_data`, since I can just use `conf.get()` to see if a past dependency has been met or not.

So it's just a syntactical complaint at this point. Maybe if I could define custom functions (e.g. `component()`) I would be able to just use keyword arguments in order to map it to the correct array index?

**Comment by nirbheek** [comment]
The reason why this hasn't been added is because nothing really needs it. It will likely be added once some meson functions start needing key-value pairs as arguments or keyword-arguments.

**Comment by haasn** [comment]
Resolved by #3490. Thanks for adding it, it's really useful.
